org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServiceAccessor.java

```diff
@@ -420,73 +433,79 @@ public static LanguageServerWrapper getLSWrapperForConnection(@NonNull IProject
 
 	private static LanguageServerWrapper getLSWrapper(@Nullable IProject project,
 			@NonNull LanguageServerDefinition serverDefinition, @Nullable IPath initialPath) throws IOException {
-		LanguageServerWrapper wrapper = null;
+
+		final Predicate<LanguageServerWrapper> serverSelector = w -> project != null && w.canOperate(project)
+				&& w.serverDefinition.equals(serverDefinition);
+
+		var matchingServer = startedServers.stream().filter(serverSelector).findFirst();
+		if (matchingServer.isPresent()) {
+			return matchingServer.get();
+		}
 
 		synchronized (startedServers) {
-			for (LanguageServerWrapper startedWrapper : getStartedLSWrappers(w -> project != null && w.canOperate(project))) {
-				if (startedWrapper.serverDefinition.equals(serverDefinition)) {
-					wrapper = startedWrapper;
-					break;
-				}
-			}
-			if (wrapper == null) {
-				wrapper = project != null ? new LanguageServerWrapper(project, serverDefinition) :
-					new LanguageServerWrapper(serverDefinition, initialPath);
-				wrapper.start();
+			// check again while holding the write lock
+			matchingServer = startedServers.stream().filter(serverSelector).findFirst();
+			if (matchingServer.isPresent()) {
+				return matchingServer.get();
 			}
 
+			final var wrapper = project != null //
+					? new LanguageServerWrapper(project, serverDefinition)
+					: new LanguageServerWrapper(serverDefinition, initialPath);
+			wrapper.start();
+
 			startedServers.add(wrapper);
+			return wrapper;
 		}
-		return wrapper;
 	}
 
 	private static LanguageServerWrapper getLSWrapperForConnection(@NonNull IDocument document,
 			@NonNull LanguageServerDefinition serverDefinition, @Nullable IPath initialPath) throws IOException {
-		LanguageServerWrapper wrapper = null;
+
+		final Predicate<LanguageServerWrapper> serverSelector = w -> w.canOperate(document);
```

Comment:
are we not missing the condition `&& w.serverDefinition.equals(serverDefinition)` ? According to my diff, it is no longer there. Or is it by intention?

Reply:
> are we not missing the condition `&& w.serverDefinition.equals(serverDefinition)` ? According to my diff, it is no longer there. Or is it by intention?

Man, you are thorough. Good catch, appreciated!!